src/core/utils/Accumulator.hpp

```diff
@@ -0,0 +1,111 @@
+#ifndef CORE_UTILS_ACCUMULATOR
+#define CORE_UTILS_ACCUMULATOR
+
+#include <boost/archive/binary_iarchive.hpp>
+#include <boost/archive/binary_oarchive.hpp>
+#include <boost/serialization/string.hpp>
+#include <boost/serialization/vector.hpp>
+#include <boost/iostreams/device/array.hpp>
+#include <boost/iostreams/stream.hpp>
+
+namespace Utils {
+
+template <typename T> struct AccumulatorData {
+  AccumulatorData()
+      : m_mean_old(0.0), m_mean_new(0.0), m_variance_old(0.0),
+        m_variance_new(0.0) {}
+  T m_mean_old;
+  T m_mean_new;
+  T m_variance_old;
+  T m_variance_new;
+private:
+  // Allow serialization to access non-public data members.
+  friend class boost::serialization::access;
+
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned version) {
+    ar & m_mean_old & m_mean_new & m_variance_old & m_variance_new;
+  }
+};
+
+class Accumulator {
+public:
+  Accumulator(std::size_t N) : m_n(0), m_acc_data(N) {}
+  void operator()(const std::vector<double>&);
+  std::vector<double> get_mean() const;
+  std::vector<double> get_variance() const;
+
+private:
+  std::size_t m_n;
+  std::vector<AccumulatorData<double>> m_acc_data;
+  // Allow serialization to access non-public data members.
+  friend class boost::serialization::access;
+
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned version) {
+    ar & m_n & m_acc_data;
+  }
+};
+
+inline void Accumulator::operator()(const std::vector<double> &data) {
+  if (data.size() != m_acc_data.size())
+    throw std::runtime_error(
+        "The given data size does not fit the initialized size!");
+  ++m_n;
+  if (m_n == 1) {
+    auto acc_data_it = m_acc_data.begin();
+    auto data_it = data.begin();
+    for (; acc_data_it != m_acc_data.end() and data_it != data.end();
+         ++acc_data_it, ++data_it) {
+      (*acc_data_it).m_mean_old = (*acc_data_it).m_mean_new = *data_it;
+      (*acc_data_it).m_variance_old = (*acc_data_it).m_variance_new = 0.0;
+    }
+  } else {
+    // Calculate mean value.
+    auto acc_data_it = m_acc_data.begin();
```

Comment:
This is  very noisy, I think you can easily be simplified by replacing this loop and the next one by one call to `std::transform`, which would also be more efficient. I think the equivalent call is

```c++
std::transform(acc_data.begin(), acc_data.end(), data.begin(), acc_data.begin(), 
    [](AccumulatorData<double> a, double d) -> AccumulatorData<double> {
      auto const new_mean = a.mean + (d - a.mean) / m_n;
      auto const new_variance = ((m_n-1) * a.variance + (d - a.mean ...
     ....

    return {new_mean, new_variance};
   });
```
This actually allows to make `AccumulatorData` immutable. Also think that you do not need to store two values. After the last loop `m_mean_old == m_mean_new`, the same for the variance.